sdk/testing/stepwise/environments/docker/environment.go

```diff
@@ -0,0 +1,851 @@
+package docker
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	mathrand "math/rand"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/go-connections/nat"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/internalshared/reloadutil"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/testing/stepwise"
+	"golang.org/x/net/http2"
+
+	docker "github.com/docker/docker/client"
+	uuid "github.com/hashicorp/go-uuid"
+)
+
+var _ stepwise.Environment = &DockerCluster{}
+
+// DockerCluster is used to managing the lifecycle of the test Vault cluster
+type DockerCluster struct {
+	// PluginName is the input from the test case
+	PluginName string
+	// ClusterName is a UUID name of the cluster.
+	ClusterName string
+
+	// MountOptions are a set of options for registering and mounting the plugin
+	MountOptions stepwise.MountOptions
+
+	RaftStorage   bool
+	CACertBytes   []byte
+	CACertPEM     []byte
+	CAKeyPEM      []byte
+	CACertPEMFile string
+	ID            string
+	ClusterName   string
+	RootCAs       *x509.CertPool
+	CACert        *x509.Certificate
+	CAKey         *ecdsa.PrivateKey
+	CleanupFunc   func()
+	SetupFunc     func()
+	ClusterNodes  []*DockerClusterNode
+
+	// networkID tracks the network ID of the created docker network
+	networkID string
+
+	barrierKeys  [][]byte
+	recoveryKeys [][]byte
+	tmpDir       string
+
+	clientAuthRequired bool
+	// the mountpath of the plugin under test
+	mountPath string
+	// rootToken is the initial root token created when the Vault cluster is
+	// created.
+	rootToken string
+}
+
+// Teardown stops all the containers.
+func (rc *DockerCluster) Teardown() error {
```

Comment:
```suggestion
func (dc *DockerCluster) Teardown() error {
```

Reply:
How come the caller is sometimes named `rc` and for others `dc` on `*DockerCluster`?

Reply:
That's a left-over from the prototype code I borrowed this from, where the cluster type was `RealCluster` and not `DockerCluster`. I renamed them all `dc` in 4e548aff7